src/vs/workbench/api/common/extHostNotebook.ts

```diff
@@ -372,6 +378,113 @@ export class ExtHostNotebookController implements ExtHostNotebookShape {
 		return fileStats;
 	}
 
+	/**
+	 * Search for query in all notebooks that can be deserialized by the serializer fetched by `handle`.
+	 *
+	 * @param handle used to get notebook serializer
+	 * @param filenamePattern the filename pattern to match files that can be deserialized
+	 * @param textQuery the text query to search using
+	 * @param token cancellation token
+	 * @returns `IRawClosedNotebookFileMatch` for every file. Files without matches will just have a `IRawClosedNotebookFileMatch`
+	 * 	with no `cellResults`. This allows the caller to know what was searched in already, even if it did not yield results.
+	 */
+	async $searchInNotebooks(handle: number, filenamePattern: string[], textQuery: ITextQuery, token: CancellationToken): Promise<{ results: IRawClosedNotebookFileMatch[]; limitHit: boolean }> {
+		const serializer = this._notebookSerializer.get(handle)?.serializer;
+		if (!serializer) {
+			return {
+				limitHit: false,
+				results: []
+			};
+		}
+
+		const runFileQueries = async (includes: (string)[], token: CancellationToken, textQuery: ITextQuery): Promise<URI[]> => {
+			const uris = new ResourceSet();
+			await Promise.all(includes.map(include => {
+				const query: IFileQuery = {
+					type: QueryType.File,
+					filePattern: include,
+					folderQueries: textQuery.folderQueries,
+					maxResults: textQuery.maxResults,
+				};
+				return this._extHostSearch.$doInternalFileSearchWithCustomCallback(query, token, (data) => {
+					data.forEach(e => {
+						uris.add(URI.revive(e));
+					});
+				});
+			}));
+
+			return Array.from(uris.keys());
+		};
+
+		const filesToScan = await runFileQueries(filenamePattern, token, textQuery);
+
+		const results = new ResourceMap<INotebookFileMatchNoModel>();
+		let limitHit = false;
+		const promises = filesToScan.map(async (uri) => {
+			const cellMatches: INotebookCellMatchNoModel[] = [];
+
+			try {
+				if (token.isCancellationRequested) {
+					return;
+				}
+				if (textQuery.maxResults && [...results.values()].reduce((acc, value) => acc + value.cellResults.length, 0) > textQuery.maxResults) {
+					limitHit = true;
+					return;
+				}
+				const notebook = await this._notebookDataCache.getNotebookData(uri, async (data: VSBuffer) => {
```

Comment:
It's mostly for performance, since the process of reading and deserializing the file might seem expensive? I should rename this, it's actually getting the text for deserialized notebooks. https://github.dev/microsoft/vscode/blob/3adeb2541c2d634f2871885a354a73c7aedb8d59/src/vs/workbench/api/common/notebookSearchExtHost.ts#L139-L141

Reply:
@andreamah `this._documents` is already in deserialized form so we don't need to convert from buffer to notebook model again.

Reply:
I just used `serializer.deserializeNotebook` in that case.